File: crypto_agent/tools/switch_exchange.py
from .registry import register_tool


@register_tool(
    name="switch_exchange",
    description="Switch the active exchange. Shows all connected exchanges if no exchange_id given.",
    schema={
        "type": "object",
        "properties": {
            "exchange_id": {"type": "string", "description": "Exchange to switch to, e.g. okx, gateio, binance"},
        },
        "required": [],
    },
)
async def handle_switch_exchange(exchange_manager, exchange_id: str = "", **_) -> str:
    try:
        exchanges = exchange_manager.list()
        active = exchange_manager.active_id

        if not exchange_id:
            lines = ["Connected Exchanges:", "=" * 35]
            for ex_id in exchanges:
                marker = " ← active" if ex_id == active else ""
                lines.append(f"  • {ex_id}{marker}")

            lines.extend(["", "Exchange Status:", "=" * 45])
            for ex_id in exchanges:
                ex = exchange_manager.get(ex_id)
                try:
                    ticker = await ex.fetch_ticker("BTC/USDT")
                    lines.append(f"  ✅ {ex_id}: online (BTC=${ticker['last']:,.2f})")
                except Exception as e:
                    lines.append(f"  ❌ {ex_id}: error ({e})")
            return "\n".join(lines)

        exchange_manager.set_active(exchange_id)
        return f"Switched active exchange to: {exchange_id}"
    except KeyError as e:
        return f"Error: {e}"
    except Exception as e:
        return f"Error in switch_exchange: {e}"

File: src/tools/switch-exchange.ts
import { registerTool } from "./registry.js";

registerTool(
  "switch_exchange",
  "Switch the active exchange. Shows all connected exchanges if no exchange_id given.",
  {
    type: "object",
    properties: {
      exchange_id: { type: "string", description: "Exchange to switch to, e.g. okx, gateio, binance" },
    },
    required: [],
  },
  ["exchange_manager", "memory", "config"],
  async ({ exchange_manager, memory, config, exchange_id = "" }) => {
    try {
      const exchanges = exchange_manager.list();
      const active = exchange_manager.activeId;

      if (!exchange_id) {
        const lines = ["Connected Exchanges:", "=".repeat(35)];
        for (const exId of exchanges) {
          const marker = exId === active ? " \u2190 active" : "";
          lines.push(`  \u2022 ${exId}${marker}`);
        }
        lines.push("", "Exchange Status:", "=".repeat(45));
        for (const exId of exchanges) {
          const ex = exchange_manager.get(exId);
          try {
            const ticker = await ex.fetchTicker("BTC/USDT");
            lines.push(`  \u2705 ${exId}: online (BTC=$${ticker.last.toLocaleString("en-US", { minimumFractionDigits: 2 })})`);
          } catch (e: any) {
            lines.push(`  \u274c ${exId}: error (${e.message ?? e})`);
          }
        }
        return lines.join("\n");
      }

      exchange_manager.setActive(exchange_id);
      memory?.setDaemonState("active_exchange", exchange_id);
      const identity = memory?.ensureDefaultIdentity?.({
        exchangeId: exchange_id,
        mode: config?.paperTrading ? "PAPER" : "LIVE",
        name: "default",
      });
      if (identity) {
        memory?.setDaemonState("active_funding_account_id", identity.fundingAccount.id);
        memory?.setDaemonState("active_trading_account_id", identity.tradingAccount.id);
        memory?.setDaemonState("active_bot_id", identity.bot.id);
      }
      return `Switched active exchange to: ${exchange_id}`;
    } catch (e: any) {
      return `Error: ${e.message ?? e}`;
    }
  },
);
